pkg/executor/sortexec/topn.go

```diff
@@ -166,76 +253,319 @@ func (e *TopNExec) loadChunksUntilTotalLimit(ctx context.Context) error {
-		newRowPtrs = append(newRowPtrs, newRowPtr)
+func (e *TopNExec) executeTopNWhenSpillTriggered(ctx context.Context) error {
+	// idx need to be set to 0 as we need to spill all data
+	e.chkHeap.idx = 0
+	err := e.spillTopNExecHeap()
+	if err != nil {
+		return err
+	}
+
+	// Wait for the finish of chunk fetcher
+	fetcherWaiter := util.WaitGroupWrapper{}
+	// Wait for the finish of all workers
+	workersWaiter := util.WaitGroupWrapper{}
+
+	// Fetch chunks from child and put chunks into chunkChannel
+	fetcherWaiter.Run(func() {
+		e.fetchChunksFromChild(ctx)
+	})
+
+	for i := range e.spillHelper.workers {
+		worker := e.spillHelper.workers[i]
+		workersWaiter.Run(func() {
+			worker.run()
+		})
 	}
-	newRowChunks.GetMemTracker().SetLabel(memory.LabelForRowChunks)
-	e.memTracker.ReplaceChild(chkHeap.rowChunks.GetMemTracker(), newRowChunks.GetMemTracker())
-	chkHeap.rowChunks = newRowChunks
 
-	e.memTracker.Consume(int64(8 * (len(newRowPtrs) - len(chkHeap.rowPtrs))))
-	chkHeap.rowPtrs = newRowPtrs
+	fetcherWaiter.Wait()
+	workersWaiter.Wait()
 	return nil
 }
+
+func (e *TopNExec) executeTopN(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(e.resultChannel, r)
+		}
+
+		close(e.resultChannel)
+	}()
+
+	heap.Init(e.chkHeap)
+	for uint64(len(e.chkHeap.rowPtrs)) > e.chkHeap.totalLimit {
+		// The number of rows we loaded may exceeds total limit, remove greatest rows by Pop.
+		heap.Pop(e.chkHeap)
+		e.chkHeap.droppedRowNum++
+	}
+
+	if err := e.executeTopNWhenNoSpillTriggered(ctx); err != nil {
+		e.resultChannel <- rowWithError{err: err}
+		return
+	}
+
+	if e.spillHelper.isSpillNeeded() {
+		if err := e.executeTopNWhenSpillTriggered(ctx); err != nil {
+			e.resultChannel <- rowWithError{err: err}
+			return
+		}
+	}
+
+	e.generateTopNResults()
+}
+
+// Return true when spill is triggered
+func (e *TopNExec) generateTopNResultsWHenNoSpillTriggered() bool {
+	rowPtrNum := len(e.chkHeap.rowPtrs)
+	for ; e.chkHeap.idx < rowPtrNum; e.chkHeap.idx++ {
+		if e.chkHeap.idx%10 == 0 && e.spillHelper.isSpillNeeded() {
+			return true
+		}
+		e.resultChannel <- rowWithError{row: e.chkHeap.rowChunks.GetRow(e.chkHeap.rowPtrs[e.chkHeap.idx])}
+	}
+	return false
+}
+
+func (e *TopNExec) generateResultWithMultiWayMerge(offset int64, limit int64) error {
+	multiWayMerge := newMultiWayMerger(&diskSource{sortedRowsInDisk: e.spillHelper.sortedRowsInDisk}, e.lessRow)
+
+	err := multiWayMerge.init()
+	if err != nil {
+		return err
+	}
+
+	outputRowNum := int64(0)
+	for {
+		if limit != -1 && outputRowNum >= limit {
+			return nil
+		}
+
+		row, err := multiWayMerge.next()
+		if err != nil {
+			return err
+		}
+
+		if row.IsEmpty() {
+			return nil
+		}
+
+		if outputRowNum >= offset {
+			select {
+			case <-e.finishCh:
+				return nil
+			case e.resultChannel <- rowWithError{row: row}:
+			}
+		}
+		outputRowNum++
+		injectParallelSortRandomFail(1)
+	}
+}
+
+func (e *TopNExec) generateTopNResultsWhenSpillTriggered() error {
```

Comment:
Seems this function does not consider the case that some rows are already sent to `resultChannel` in `generateTopNResultsWHenNoSpillTriggered` before spill is triggered.

Reply:
> Seems this function does not consider the case that some rows are already sent to `resultChannel` in `generateTopNResultsWHenNoSpillTriggered` before spill is triggered.

We have considered this case. We denote rows that have been sent to `resultChannel` as A. When some rows have been sent to `resultChannel`, only rows that after A will be spilled to disk, and we will send all rows that spilled to disk to the `resultChannel` in restoring stage.